src/buffer/out/textBuffer.cpp

```diff
@@ -1928,90 +1928,73 @@ void TextBuffer::_ExpandTextRow(til::inclusive_rect& textRow) const
     data.text.reserve(rows);
-    if (copyTextColor)
+    if (copyAttribute)
     {
-        data.FgAttr.reserve(rows);
-        data.BkAttr.reserve(rows);
+        data.attrs.reserve(rows);
     }
 
-    // for each row in the selection
     for (size_t i = 0; i < rows; i++)
     {
-        const auto iRow = selectionRects.at(i).top;
-
-        const auto highlight = Viewport::FromInclusive(selectionRects.at(i));
-
-        // retrieve the data from the screen buffer
-        auto it = GetCellDataAt(highlight.Origin(), highlight);
-
-        // allocate a string buffer
         std::wstring selectionText;
-        std::vector<COLORREF> selectionFgAttr;
-        std::vector<COLORREF> selectionBkAttr;
+        til::small_rle<TextAttribute, uint16_t, 1> attrs;
 
-        // preallocate to avoid reallocs
-        selectionText.reserve(gsl::narrow<size_t>(highlight.Width()) + 2); // + 2 for \r\n if we munged it
-        if (copyTextColor)
-        {
-            selectionFgAttr.reserve(gsl::narrow<size_t>(highlight.Width()) + 2);
-            selectionBkAttr.reserve(gsl::narrow<size_t>(highlight.Width()) + 2);
-        }
+        const auto iRow = selectionRects.at(i).top;
+        const auto& row = GetRowByOffset(iRow);
+        const auto colBegin = gsl::narrow_cast<uint16_t>(selectionRects.at(i).left);
+        auto colEnd = gsl::narrow_cast<uint16_t>(selectionRects.at(i).right + 1); // end column exclusive
 
-        // copy char data into the string buffer, skipping trailing bytes
-        while (it)
+        // We apply formatting to rows if the row was NOT wrapped or formatting
+        // of wrapped rows is allowed.
+        const auto shouldFormatRow = formatWrappedRows || !row.WasWrapForced();
+
+        if (trimTrailingWhitespace && shouldFormatRow)
         {
-            const auto& cell = *it;
+            // update end column to exclude trailing whitespace
 
-            if (cell.DbcsAttr() != DbcsAttribute::Trailing)
+            const auto text = row.GetText();
+            const auto begin = text.rbegin();
+            const auto end = text.rend();
+            auto it = begin;
+            for (; it != end; ++it)
             {
-                const auto chars = cell.Chars();
-                selectionText.append(chars);
-
-                if (copyTextColor)
+                if (*it != L' ')
                 {
-                    const auto cellData = cell.TextAttr();
-                    const auto [CellFgAttr, CellBkAttr] = GetAttributeColors(cellData);
-                    for (size_t j = 0; j < chars.size(); ++j)
-                    {
-                        selectionFgAttr.push_back(CellFgAttr);
-                        selectionBkAttr.push_back(CellBkAttr);
-                    }
+                    break;
                 }
             }
 
-            ++it;
+            const auto textTrimmedLength = gsl::narrow_cast<uint16_t>(end - it);
+            colEnd = std::min(colEnd, textTrimmedLength);

```

Comment:
I don't think the `textTrimmedLength` code is correct... `end - it` returns the number of trailing whitespaces so `colEnd` should be `row.size() - (end - it)`, no? Also, we can simplify this using either `find_last_not_of(L' ')` or `ROW::MeasureRight()`. That is:
```cpp
colEnd = std::min(colEnd, row.MeasureRight());
```

Reply:
> `row.size() - (end - it)`

That would be correct if were to use `.begin()`/`.end()`, but here we are using `.rbegin()`/`.rend()`, isn't?

> Also, we can simplify this using either find_last_not_of(L' ') or ROW::MeasureRight()

I faced some problem when I tried to use  `ROW::MeasureRight()` for getting the trimmed length. IIRC, the row with the cursor was always `_wrapForced = true`, which made `MeasureRight()` return `_columnCount`. However, it seems to work now so I guess I'll update this code to use `MeasureRight()` 🙂

Reply:
Forgot to mention that `MeasureRight()` is better in the sense that it doesn't trim trailing whitespaces of wrapped rows. This is because, technically, the trailing whitespaces on a wrapped row are not considered trailing whitespace of the line that was wrapped.

Reply:
Oh right, `end - it` returns the number of characters from the start to the iterator. But that's also wrong, since there isn't a 1:1 mapping between the number of characters and columns.
I'm glad that `MeasureRight()` works for you now!